### PR TITLE
Support for out-of-order validator signatures

### PR DESCRIPTION
**Pull request by cyl19970726**

### CODE REVIEW

File: contracts/lib/BlockDecoder.sol [hunk 1]
```diff
@@ -169,7 +170,11 @@ library BlockDecoder {
                 require(vaddr == validators[i], "validator no exist");
                 idx = i;
             } else {
-                assert(false);
+                bool exist;
+                (exist, idx) = _validatorIndex(vaddr, validators);
```
Comment (qizhou) [inline contracts/lib/BlockDecoder.sol]: Need a way to make sure the idx is not used twice?  Perhaps a bitmap array may help.

Comment (cyl19970726) [reply contracts/lib/BlockDecoder.sol]: have solved this issue

File: contracts/lib/BlockDecoder.sol [hunk 2]
```diff
@@ -169,7 +170,11 @@ library BlockDecoder {
...
+                if (!exist) {
+                    continue;
+                }
             }
 
             bytes memory signMsg = voteSignBytes(commit, chainId, i);
```
Comment (qizhou) [inline contracts/lib/BlockDecoder.sol:185]: Should `i` be replaced by `idx`?

File: contracts/lib/BlockDecoder.sol [hunk 3]
```diff
@@ -190,6 +195,25 @@ library BlockDecoder {
...
     }
 
+    function _validatorIndex(address val, address[] memory vals) internal pure returns (bool exist, uint256 index) {
+        for (index = 0; index < vals.length; index++) {
+            if (val == vals[index]) {
+                exist = true;
+                if (index != vals.length - 1) {
+                    vals[index] = vals[vals.length - 1];
+                }
+                uint256 newLen = vals.length - 1;
```
Comment (qizhou) [inline contracts/lib/BlockDecoder.sol]: voting power should be changed accordingly?

File: contracts/lib/BlockDecoder.sol [hunk 4]
```diff
@@ -195,23 +196,30 @@ library BlockDecoder {
...
+                    powers[actualLen - 1] = tmpPower;
                 }
 
                 break;
             }
         }
 
-        return (exist, index);
+        if (exist){
+            return (exist, index,actualLen-1);
```
Comment (qizhou) [inline contracts/lib/BlockDecoder.sol]: Should `index` here be `actualLen - 1`?

**Review by qizhou — COMMENTED** [review]
Looks like the code needs to be prettier?

File: contracts/lib/BlockDecoder.sol (hunk 4 above)
Comment (qizhou) [reply contracts/lib/BlockDecoder.sol]: I think we should add a test to cover the issue?

**Comment by cyl19970726** [comment]
> Looks like the code needs to be prettier?

done

**Comment by qizhou** [comment]
Any tests to cover the issues found in recent reviews?

**Comment by cyl19970726** [comment]
> Any tests to cover the issues found in recent reviews?

All issues have been covered by tests

File: contracts/lib/BlockDecoder.sol [hunk 5]
```diff
@@ -162,19 +163,29 @@ library BlockDecoder {
         require(votePowers.length == validators.length, "incorrect length");
         uint256 talliedVotingPower;
         uint256 idx;
+        uint256 actualLen = validators.length;
         for (uint256 i = 0; i < commit.Signatures.length; i++) {
-            address vaddr = commit.Signatures[i].ValidatorAddress;
+            // address vaddr = commit.Signatures[i].ValidatorAddress;
```
Comment (qizhou) [inline contracts/lib/BlockDecoder.sol]: remove unused line?

**Review by qizhou — APPROVED** [review]

**Pull request merged**